{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Project Description:\n",
      "----------------------\n",
      "\n",
      "### Please refer the end part of the documentation for instructions to run the program, with the files submitted.\n",
      "### Also, these files have been submitted in github at this url: \n",
      "#### Github url: https://github.com/kannanganat/Final-Project.git\n",
      "\n",
      "- Original site for yeast related data: http://www.yeastgenome.org/\n",
      "- site for data download: http://yeastmine.yeastgenome.org/yeastmine/begin.do \n",
      "- site for software documentation: https://intermine.readthedocs.org/en/latest/\n",
      "\n",
      "\n",
      "- Yeastmine is a dataminning tool that was developed as a result of the collaboration of SGD (Saccaharomyces Genome Database) and the intermine project at the Cambridge Systems Biology Center\n",
      "\n",
      "\n",
      "- One could use this site to generate specific query results of the vast amount of data in the SGD website.\n",
      "\n",
      "> - The initial phase of my project will download a specific set of information from SGD via the yeastmine webpage.\n",
      "> - This information will be used to generate multiple graphs (depending on the biological questions that is being pursued)\n",
      "    - for plotting locations of ribosomal protein coding genes on 16 chromosomes in yeast\n",
      "    - for plotting interacting proteins per each of the ribosomal protein paralogs\n",
      "    This will be done using networkx.\n",
      "    \n",
      "> - These graphs and the related code will be published on a webpage - using django. \n",
      "    - The graph for proteins interacting with ribosomal proteins will be set up in such a manner that one could look at each ribosomal proteins and its inteactors individually. There will be a dropdown menu in the django website that can be used for this purpose.\n",
      "    \n",
      "> - Details on the interaction graph:\n",
      "    - Off all the interactions that a ribosomal protein has, I thought that it would interesting to look at those that are manually curated in the SDG website. Each ribosomal protein will form the center node and the interactors will be plotted as other nodes. The edges will contain information on the type of experiment used for establishing each interaction. \n",
      "    \n",
      "\n",
      "### 1) How to install the package\n",
      "\n",
      "- In order to connect to the yeastmine webpage, one must install python webservice client : intermine\n",
      "> using \"sudo easy_install intermine\"\n",
      "- networkx and django - as previously demonstrated in lecture notes. \n",
      "\n",
      "\n",
      "### 2) List any dependencies\n",
      "\n",
      "- python version 2.6 or newer\n",
      "- pdoc - for documentation\n",
      "- Intermine webservice client\n",
      "- matplotlib\n",
      "- networkx\n",
      "- django \n",
      "\n",
      "### 3) Give examples of how to run the program\n",
      "\n",
      "- One can run the downloaded script directly in ipython notebook or in the terminal using \"python path/to/downloaded/script.py\"\n",
      "- before running any script one must run the following command:to connect to the website and must be written with every code\n",
      "\n",
      "##### from intermine.webservice import Service\n",
      "##### service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "\n",
      "#### One example of getting a query: \n",
      "\n",
      "-# Get a new query on the class (table) you will be querying:\n",
      "-query = service.new_query(\"GeneFlankingRegion\")\n",
      "\n",
      "-# Type constraints should come early - before all mentions of the paths they constrain\n",
      "-#query.add_constraint(\"overlappingFeatures\", \"Gene\")\n",
      "\n",
      "-# The view specifies the output columns\n",
      "-query.add_view(\n",
      "-    \"gene.secondaryIdentifier\", \"gene.symbol\", \"direction\", \"length\",\n",
      "-    \"gene.sequence.residues\"\n",
      "-)\n",
      "\n",
      "-# You can edit the constraint values below\n",
      "-query.add_constraint(\"organism.shortName\", \"=\", \"S. cerevisiae\", code = \"C\")\n",
      "-query.add_constraint(\"gene\", \"LOOKUP\", \"RPL10\", \"S. cerevisiae\", code = \"B\")\n",
      "-query.add_constraint(\"direction\", \"=\", \"upstream\", code = \"D\")\n",
      "-#query.add_constraint(\"overlappingFeatures.featureType\", \"=\", \"ORF\", code = \"E\")\n",
      "-query.add_constraint(\"length\", \"=\", \"2000\", code = \"A\")\n",
      "\n",
      "-# Uncomment and edit the code below to specify your own custom logic:\n",
      "-# query.set_logic(\"B and D and E and A and C\")\n",
      "\n",
      "-for row in query.rows():\n",
      "-    print row[\"gene.secondaryIdentifier\"], row[\"gene.symbol\"], row[\"direction\"], row[\"length\"], \\\n",
      "-        row[\"gene.sequence.residues\"]\n",
      "\n",
      "#### output for this script will be as follows:\n",
      "\n",
      "-YLR075W RPL10 upstream 2000 ATGGCTAGAAGACCAGCTAGATGTTACAGATACCAAAAGAACAAGCCTTACCCAAAGTCTAGATACAACAGAGCTGTTCCAGACTCCAAGATCAGAATCTACGATTTGGGTAAGAAGAAGGCTACCGTCGATGAATTCCCATTGTGTGTCCATTTGGTTTCCAACGAATTGGAACAATTGTCTTCTGAAGCTTTGGAAGCTGCTCGTATCTGTGCCAACAAGTACATGACTACTGTCTCCGGTAGAGATGCTTTCCACTTAAGAGTCAGAGTCCATCCTTTCCATGTCTTGAGAATCAACAAGATGTTGTCTTGTGCCGGTGCGGATAGATTGCAACAAGGTATGAGAGGTGCTTGGGGTAAGCCACACGGTTTGGCCGCTCGTGTCGACATTGGTCAAATTATCTTCTCTGTCAGAACCAAGGACAGCAACAAGGATGTTGTCGTTGAAGGTTTGAGAAGAGCCAGATACAAGTTCCCAGGTCAACAAAAGATTATTTTGTCTAAGAAGTGGGGTTTCACCAACTTGGACAGACCAGAATACTTGAAGAAGAGAGAAGCTGGTGAAGTCAAGGACGACGGTGCTTTCGTTAAGTTCTTGTCCAAGAAGGGTTCTTTGGAAAACAACATCAGAGAATTCCCAGAATACTTTGCTGCTCAAGCTTAA\n",
      "\n",
      "#### Graphs will be made using matplotlib and networkx"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Actual Project related codes and detailed documentation of each step"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Files required for running the code: submitted as seperate files\n",
      "#### also, a snap shot of the execution is also submitted. \n",
      "> grpahs.py\n",
      "\n",
      "> interactions.py\n",
      "\n",
      "> url.py\n",
      "\n",
      "> views.py"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### Code used for generating the graph - ribosomal protein coding genes plotted on the yeast genome"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### please note that this code will not run by itself.\n",
      "#### It is written in such a way that the code will create a graph when run through Django."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# code for generating the graph - for all ribosomal protin coding genes\n",
      "# please note that this code will not run by itself. \n",
      "# It is written in such a way that the code will create a graph when run through Django.\n",
      "\n",
      "#!/usr/bin/env python\n",
      "import matplotlib\n",
      "matplotlib.use('SVG')\n",
      "from intermine.webservice import Service\n",
      "from matplotlib import pyplot as plt\n",
      "#import random\n",
      "\n",
      "# Importing the necessary modules for the code to run\n",
      "\n",
      "def generate_graph():\n",
      "\"\"\" This code will help retrieve the ribosomal protein coding genes (symbol), their chromosome locations\n",
      "(the start and end positions of a gene), then plot the positions of each gene on the graph in a svg file. \n",
      "\"\"\"\n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "    \n",
      "    # Get a new query on the class (table) you will be querying:\n",
      "    query = service.new_query(\"Gene\")\n",
      "    \n",
      "    # Type constraints should come early - before all mentions of the paths they constrain\n",
      "    # this GO term was used to narrow down the ribosomal protein coding genes from the whole geneome.\n",
      "    query.add_constraint(\"goAnnotation.ontologyTerm\", \"GOTerm\")\n",
      "    \n",
      "    # The view specifies the output columns\n",
      "    # we are interested mainly in the gene name \"symbol\", chromosome number \"chromosome.primaryIdentifier\"\n",
      "    # chromosome locations \"chromosomeLocation.end\" and \"chromosomeLocation.start\"\n",
      "    # \"chromosome.length\" - will be used to draw the whole chromosome. \n",
      "    # The positions of each gene plotted are an average of the start and end points of a gene.\n",
      "    query.add_view(\n",
      "        \"secondaryIdentifier\", \"symbol\", \"goAnnotation.ontologyTerm.identifier\",\n",
      "        \"description\", \"chromosome.primaryIdentifier\", \"chromosomeLocation.start\",\n",
      "        \"chromosomeLocation.end\", \"chromosome.length\"\n",
      "    )\n",
      "    \n",
      "    # This query's custom sort order is specified below:\n",
      "    query.add_sort_order(\"Gene.symbol\", \"ASC\")\n",
      "\n",
      "    # You can edit the constraint values below\n",
      "    query.add_constraint(\"goAnnotation.qualifier\", \"IS NULL\", code = \"C\")\n",
      "    query.add_constraint(\"goAnnotation.qualifier\", \"!=\", \"NOT\", code = \"B\")\n",
      "    query.add_constraint(\"goAnnotation.ontologyTerm.name\", \"=\", \"cytoplasmic translation\", code = \"A\")\n",
      "    query.add_constraint(\"name\", \"ONE OF\", [\"Ribosomal Protein of the Large subunit\", \"Ribosomal Protein of the Small subunit\"], code = \"D\")\n",
      "\n",
      "    # Your custom constraint logic is specified with the code below:\n",
      "    query.set_logic(\"A and (B or C) and D\")\n",
      "\n",
      "\n",
      "    chromosome={}\n",
      "    for row in query.rows(): # has all the data\n",
      "# using this for loop, we are iterating through the list generated by the code above.\n",
      "# first we check if the chromosome identifier (number), is there in the dictionary called chromosome\n",
      "# if not, then we add that chromosome number as the key and its corresponding value will be another dictionary.\n",
      "# This sub-dictionary will contain the length of the chromosome as a key and the \"genes\" as its value\n",
      "# \"genes\" will carry the information on the gene name \"symbol\" and its position. \n",
      "        if row[\"chromosome.primaryIdentifier\"] not in chromosome.keys():\n",
      "            chromosome[row[\"chromosome.primaryIdentifier\"]] = {\"length\":row[\"chromosome.length\"],\"genes\":[]}\n",
      "            \n",
      "        chromosome[row[\"chromosome.primaryIdentifier\"]][\"genes\"].append({'symbol':row[\"symbol\"],'start': row[\"chromosomeLocation.start\"], 'end':row[\"chromosomeLocation.end\"]})\n",
      "    \n",
      "\n",
      "    all_chr_ids = chromosome.keys()\n",
      "    all_chr_length = []\n",
      "    for chr in all_chr_ids:\n",
      "# for plotting the graph, an example graph in the matplotlib website was used:\n",
      "# http://matplotlib.org/examples/lines_bars_and_markers/barh_demo.html\n",
      "# this primary code was modified to suit our needs. \n",
      "        all_chr_length.append( chromosome[chr]['length'] )\n",
      "    fig = plt.figure()\n",
      "    fig.set_size_inches(25,10)\n",
      "    y_pos = range(len(all_chr_ids))\n",
      "    plt.barh(y_pos, all_chr_length, align='center', alpha=0.4)  #draws the horizontal bar graph from every element in 'y-pos'(y-axis) to 'all_chr_length'(x-axis)\n",
      "    plt.yticks(y_pos,all_chr_ids ) #Says position and label for y-axes\n",
      "\n",
      "    plt.xlabel('Gene Positions') #sets label for x-axes\n",
      "    plt.title('Ribosomal protein Genes in Yeast Genome')\n",
      "\n",
      "    chromosome_y_axis = 0\n",
      "    for chr in all_chr_ids: # just to get the average value for gene positions, to be used for plotting. \n",
      "        for gene in chromosome[chr]['genes']:\n",
      "            posn = (gene['start']+gene['end'])/2\n",
      "            plt.plot(posn,chromosome_y_axis,'ro')\n",
      "            plt.annotate(gene['symbol'],(posn-2,chromosome_y_axis+0.1),rotation='vertical')\n",
      "        chromosome_y_axis+= 1    \n",
      "        \n",
      "    fig.savefig('static/rp_positions.svg')\n",
      "# finally the graph is saved as an svg file. This file will be used by Django while creating the website.     \n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## code used for creating the interactions graph using networkx"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#!/usr/bin/env python\n",
      "import matplotlib\n",
      "matplotlib.use('SVG')\n",
      "import networkx as nx\n",
      "from matplotlib import pyplot as plt\n",
      "from intermine.webservice import Service\n",
      "\n",
      "# Importing the necessary modules for the code to run\n",
      "\n",
      "def getInteractions():\n",
      "\n",
      "\"\"\" This code will help retrieve the ribosomal protein coding genes (symbol), the genes (gene 2) they are known to interact with\n",
      "(but only those that are manually curated in the SGD website), the experiment type used to confirm interaction\n",
      "and a brief description of the function of gene 2. \n",
      "Also, this will create a dictionary withtin a dictionary model to assign the experiment type and gene 2s for each ribosomal protein coding gene\n",
      "\"\"\"\n",
      "    service = Service(\"http://yeastmine.yeastgenome.org/yeastmine/service\")\n",
      "\n",
      "    # Get a new query on the class (table) you will be querying:\n",
      "    query = service.new_query(\"Gene\")\n",
      "\n",
      "    # Type constraints should come early - before all mentions of the paths they constrain\n",
      "    query.add_constraint(\"goAnnotation.ontologyTerm\", \"GOTerm\")\n",
      "\n",
      "    # The view specifies the output columns\n",
      "    query.add_view(\n",
      "        \"symbol\", \"interactions.details.experimentType\",\n",
      "        \"interactions.gene2.symbol\", \"interactions.gene2.briefDescription\"\n",
      "    )\n",
      "\n",
      "    # You can edit the constraint values below\n",
      "    query.add_constraint(\"goAnnotation.qualifier\", \"IS NULL\", code = \"C\")\n",
      "    query.add_constraint(\"goAnnotation.qualifier\", \"!=\", \"NOT\", code = \"B\")\n",
      "    query.add_constraint(\"goAnnotation.ontologyTerm.name\", \"=\", \"cytoplasmic translation\", code = \"A\")\n",
      "    query.add_constraint(\"name\", \"ONE OF\", [\"Ribosomal Protein of the Large subunit\", \"Ribosomal Protein of the Small subunit\"], code = \"D\")\n",
      "    query.add_constraint(\"interactions.details.annotationType\", \"=\", \"manually curated\", code = \"E\")\n",
      "\n",
      "    # Your custom constraint logic is specified with the code below:\n",
      "    query.set_logic(\"A and (B or C) and E and D\")\n",
      "\n",
      "    \n",
      "    interactions = {}\n",
      "    \n",
      "    for row in query.rows(): # the value for the key (symbol/gene) is actually a list which has each element as a dictionary\n",
      "        if row[\"symbol\"] not in interactions.keys():\n",
      "            interactions[row[\"symbol\"]] = [{ \"expt\" : row[\"interactions.details.experimentType\"], \"gene2\": row[\"interactions.gene2.symbol\"],\"desc\":row[\"interactions.gene2.briefDescription\"]}]\n",
      "        else:\n",
      "            interactions[row[\"symbol\"]].append({ \"expt\": row[\"interactions.details.experimentType\"], \"gene2\": row[\"interactions.gene2.symbol\"],\"desc\":row[\"interactions.gene2.briefDescription\"]})\n",
      "    return interactions\n",
      "\n",
      "def get_interaction(interactions,gene):\n",
      "    \n",
      "\"\"\" Using networkx a graph will be made that will demonstrate the interactors as nodes,\n",
      "the experiment type will be mentioned within the edge and a brief description of the function of gene 2 will be below the node for gene 2\n",
      "\"\"\"\n",
      "    G=nx.Graph()\n",
      "    fig = plt.figure() # what ever that is drawn using the below code, will go into the \"figure\"\n",
      "    fig.set_size_inches(25,10) # size of the figure\n",
      "    labels = {} # a dictionary to contain the brief description of gene 2\n",
      "    edge_labels = {} # a dictionary to contain the experiment types\n",
      "    labels[gene]=\"\"\n",
      "    colors = [] # empty list - to be used lated - just to differentiate between the ribosomal protein (gene1) and other gene 2s\n",
      "    for v in interactions[gene]:\n",
      "        G.add_edge(gene,v[\"gene2\"]) # adding edge, this will ensure that nodes are created automatically\n",
      "        labels[v[\"gene2\"]] = v[\"desc\"] # storing the description for each gene \n",
      "        edge_labels[(gene,v[\"gene2\"])] = v[\"expt\"] # since edges are tuples\n",
      "            \n",
      "    layout= nx.spring_layout(G,iterations=100) # reference from networkx gallery\n",
      "    layout1 = dict((n,(d[0,],d[1,]+0.08)) for n,d in layout.items()) \n",
      "# just to ensure that the description of the gene doesn't overlap with the gene name\n",
      "    \n",
      "    for n in G.nodes(): # just for coloring the gene 1 and other gene 2s\n",
      "        if n == gene:\n",
      "            colors.append('c')\n",
      "        else:\n",
      "            colors.append('r')\n",
      "        \n",
      "    nx.draw(G,layout,node_size=2500,node_color=colors,edge_color='b',style='dashed') # graphs with nodes and edges\n",
      "    nx.draw_networkx_labels(G,layout1,labels=labels,font_color='g') # node labels for gene 2s\n",
      "    nx.draw_networkx_edge_labels(G,layout,edge_labels=edge_labels,font_color='y',font_size=14) # drawing the experiment types (edge labels)\n",
      "    fig.savefig('static/rp_interactions.svg')\n",
      "# finally the graph generated through networkx will be saved as a .svg file which will be used by Django. \n",
      "    \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Basic Steps for using Django"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### initial testing of Django for the ribosomal protein coding genes graph on the yeast genome.\n",
      "\n",
      "> step 1: django-admin.py startproject project\n",
      "\n",
      "> step 2: cd project \n",
      "\n",
      "> step 3: python manage.py startapp genemap\n",
      "\n",
      "> step 4: mkdir static\n",
      "\n",
      "> step 5: Open project/settings.py\n",
      "\n",
      "\n",
      "-  and add following lines, please note to remove the bullet point or \"-\" while copying from ipython notebook. \n",
      "-  --------------------------\n",
      "\n",
      "-  STATICFILES_DIRS = (\n",
      "-    os.path.join(BASE_DIR, \"static\"),\n",
      "- )\n",
      "\n",
      "-  --------------------------\n",
      "\n",
      "> step 6: mv graphs.py interactions.py urls.py views.py genemap/ - just make sure that these files are in the genemap folder\n",
      "#### basically moving the required filed into the genemap folder.\n",
      "\n",
      "> step 7: mv index.html static/\n",
      "\n",
      "> step 8: Open project/urls.py add the following line, near a similar line (that resembles the below line)\n",
      "\n",
      "-        url(r'^genemap/',include('genemap.urls')),\n",
      "\n",
      "> step 9: python manage.py runserver\n",
      "\n",
      "> step 10: Go to following URL in a browser(tested with firefox):\n",
      "        \n",
      "        \n",
      "        "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Final Run of the program\n",
      "\n",
      "> step 1 - open terminal \n",
      "\n",
      "> step 2 - \"cd\" to the folder \"project\" - in my case it would be \"/Users/arunkannanganat/codes/project\"\n",
      "\n",
      "> step 3 - run the following command \"python manage.py runserver\"\n",
      "\n",
      "> step 4: check out the url below for the graphs\n",
      "    \n",
      "### http://127.0.0.1:8000/genemap/index  \n"
      "#### I wish to express my sincere thanks and gratitude to my friend, sri who helped me survive python \n"
      "#### I also wish to thank the TAs and Dr. Chapman and my classmates for all their help\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}
